src/client/gfs_client.cc

```diff
@@ -1,32 +1,80 @@
 #include "src/client/gfs_client.h"
+#include "src/client/client_impl.h"
+#include "src/common/utils.h"
 
 namespace gfs {
 namespace client {
 
-google::protobuf::util::Status open(const char* pathname, unsigned int flags) {
-  // Step 1. Check and validate the flags, e.g. it doesn't make sense to 
-  // open with both read and write flags
+// A singleton object that serves as the interface for the impl code to access
+// and issue calls to internal units such as cache manager, service clients to
+// master and chunk servers
+static ClientImpl* client_impl_ = nullptr;
 
-  // Creation mode
-  if (flags | OpenFlag::Create) {
+// Validate a given open flag, currently we only support four modes, namely
+// read, write, create and create | write
+bool ValidateOpenFlag(unsigned int flags) {
+  return flags == OpenFlag::Read || flags == OpenFlag::Write ||
+         flags == OpenFlag::Create || 
+         flags == (OpenFlag::Write | OpenFlag::Create);
+}
+
+google::protobuf::util::Status init_client(const std::string& config_filename,
+  const std::string& master_name, const bool resolve_hostname) {
+  if (client_impl_) {
+    return  google::protobuf::util::Status(
+                google::protobuf::util::error::ALREADY_EXISTS,
+                "ClientImpl has already been initialized successfully");
+  }
+
+  // Instantiate the client impl object
+  auto client_impl_or(ClientImpl::ConstructClientImpl(config_filename, 
+                          master_name, resolve_hostname));
+  if (!client_impl_or.ok()) {
+    return client_impl_or.status();
+  }
+
+  client_impl_ = client_impl_or.ValueOrDie();
+  return google::protobuf::util::Status::OK;
+}
+
+google::protobuf::util::Status open(const char* filename, unsigned int flags) {
+  // Make sure that init_client is called as a pre-condition
+  if (!client_impl_) {
+    return  google::protobuf::util::Status(
+                google::protobuf::util::error::FAILED_PRECONDITION,
+                "init_client must be called before calling client APIs");
+  }
+   
+  // Check and validate the flags, e.g. it doesn't make sense to 
+  // open with both read and write flags
+  if (!ValidateOpenFlag(flags)) {
+    return google::protobuf::util::Status(
+               google::protobuf::util::error::INVALID_ARGUMENT,
+               "Invalid open flag : " + std::to_string(flags)); 
   }
 
-  // Read mode
-  if (flags | OpenFlag::Read) {
+  // Check and validate the filename
+  auto check_filename_status(common::utils::CheckFilenameValidity(filename));
+  if (!check_filename_status.ok()) {
+    return check_filename_status;
   }
 
-  // Write mode
-  if (flags | OpenFlag::Write) {
+  // Creation mode
+  if (flags | OpenFlag::Create) {
```

Comment:
1. can we not use bit manipulation. it's very hard to tell exactly what this returns. consider using explict equal sign or something else?
2. how about write mode with create? shouldn't we also check for file existence and create as necessary

Reply:
I am going to leave a TODO in this area just for now, as we have some other priorities to push forward.

Reply:
As I am going over this comment again, I see what you are asking. The write + create mode is included in the above condition, because the flag for that would be (OpenFlag::Write | Open::Create). I guess writing explicit equal would mean we have to use two conditions here for single create and create + write respectively. ButI am adding a comment clarifying what the condition means.